core/src/OpenACC/Kokkos_OpenACC_Atomic.hpp

```diff
@@ -0,0 +1,1964 @@
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_OPENACC_ATOMIC_HPP
+#define KOKKOS_OPENACC_ATOMIC_HPP
+
+#include <impl/Kokkos_Atomic_Memory_Order.hpp>
+#include <impl/Kokkos_Memory_Fence.hpp>
+#include <openacc.h>
+#include <stdio.h>
+#include <algorithm>
+
+//#define KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS_ERROR
+
+#if defined(KOKKOS_ENABLE_OPENACC_ATOMICS)
+namespace Kokkos {
+// OpenACC can do:
+// Operations: +, -, *, /, &, |, ^, <<, >>
+// Type: int, unsigned int, unsigned long long int
+// Variants:   
+// atomic_exchange/compare_exchange/assign/fetch_add/fetch_sub/fetch_mul/fetch_div/fetch_mod/fetch_max/fetch_min/fetch_and/fetch_or/fetch_xor/fetch_lshift/fetch_rshift/add_fetch/sub_fetch/mul_fetch/div_fetch/mod_fetch/max_fetch/min_fetch/and_fetch/or_fetch/xor_fetch/lshift_fetch/rshift_fetch/
+// _atomic_load/store
+
+// atomic_exchange -------------------------------------------------------------
+
+#pragma acc routine seq
+template <typename T>
+__inline__ T atomic_exchange(
+    volatile T *const dest,
+    const T &val) {
+  T retval;
+  T *ptr = const_cast<T *>(dest);
+//  #pragma acc atomic capture
```

Comment:
That code was for unsupported atomic data types; I changed the code so that for unsupported atomic data types, it will abort if running on a device but it will sequentially execute the construct (without atomicity) if running on a host.